library/src/main/java/com/google/android/exoplayer2/metadata/id3/Id3Decoder.java

```diff
@@ -444,6 +450,112 @@ private static TextInformationFrame decodeTextInformationFrame(ParsableByteArray
+
+    int descriptionEndIndex = indexOfEos(data, 0, encoding);
+    String description = new String(data, 0, descriptionEndIndex, charset);
+
+    int urlStartIndex = descriptionEndIndex + 1;
+    int urlEndIndex = indexOfEos(data, urlStartIndex, encoding);
+    int urlLength = urlEndIndex - urlStartIndex;
+    String url = new String(data, urlStartIndex, urlLength, charset);
+
+    return new UrlLinkFrame(description, url);
+  }
+
+  private static ChapterFrame decodeChapterFrame(ParsableByteArray id3Data, int frameSize,
+      int majorVersion, boolean unsignedIntFrameSizeHack) throws UnsupportedEncodingException {
+    byte[] frameBytes = new byte[frameSize];
+    id3Data.readBytes(frameBytes, 0, frameSize - 1);
+
+    ParsableByteArray chapterData = new ParsableByteArray(frameBytes);
+
+    int chapterIdEndIndex = indexOfZeroByte(frameBytes, 0) + 1;
+    String chapterId = chapterData.readNullTerminatedString(chapterIdEndIndex);
+
+    chapterData.setPosition(chapterIdEndIndex);
+    int startTime = chapterData.readInt();
+    int endTime = chapterData.readInt();
+    int startOffset = chapterData.readInt();
+    int endOffset = chapterData.readInt();
+
+    String title = null;
+    String url = null;
+    ApicFrame image = null;
+
+    int frameHeaderSize = majorVersion == 2 ? 6 : 10;
+    while (chapterData.bytesLeft() >= frameHeaderSize) {
+      Id3Frame frame = decodeFrame(majorVersion, chapterData, unsignedIntFrameSizeHack);
+      if (frame == null) {
+        continue;
+      }
+      if (frame instanceof TextInformationFrame) {
+        TextInformationFrame textFrame = (TextInformationFrame)frame;
+        if (textFrame.id != null && textFrame.id.equals("TIT2")) {
+          title = textFrame.description;
+        }
+      }
+      else if (frame instanceof UrlLinkFrame) {
+        UrlLinkFrame linkFrame = (UrlLinkFrame)frame;
+        url = linkFrame.url;
+      }
+      else if (frame instanceof ApicFrame) {
+        image = (ApicFrame)frame;
+      }
+    }
+
+    return new ChapterFrame(chapterId, startTime, endTime, startOffset, endOffset, title, url, image);
+  }
+
+  private static ChapterTOCFrame decodeChapterTOCFrame(ParsableByteArray id3Data, int frameSize,
+    int majorVersion, boolean unsignedIntFrameSizeHack) throws UnsupportedEncodingException {
+    byte[] frameBytes = new byte[frameSize];
+    id3Data.readBytes(frameBytes, 0, frameSize - 1);
+
+    ParsableByteArray tocData = new ParsableByteArray(frameBytes);
+
+    int idEndIndex = indexOfZeroByte(frameBytes, 0) + 1;
+    String id = tocData.readNullTerminatedString(idEndIndex);
+    tocData.setPosition(idEndIndex);
+
+    int flags = tocData.readUnsignedByte();
+    boolean isRoot = (flags & 0x0002) != 0;
+    boolean isOrdered = (flags & 0x0001) != 0;
+
+    int entryCount = tocData.readUnsignedByte();
+    String[] children = new String[entryCount];
+    for (int i = 0; i < entryCount && tocData.bytesLeft() > 0; i++) {
```

Comment:
Is the tocData.bytesLeft() > 0 check really necessary? i < entryCount should be sufficient to ensure this unless the metadata is really quite broken.